### PR TITLE
Feat/training

### PR DESCRIPTION
**Pull request by deep-diver**
Additional notebook to handle TFRecords in `ImportExampleGen` is included. It works on my local machine which is equipped with lots of memory, but it does not work on Vertex AI. The notebook can be tested on a small local machine to reproduce the problem.

For now, only the `train-00-*.tfrec` and `val-00-*.tfrec` are included as the input dataset.

Some brief intro for each files under `training_pipeline`: 
- `pipeline/configs.py`: set of variables for pipeline name, vertex ai training config, gcp project id and region
- `models/model.py`: modeling code
- `local_runner.py`: a runner to run pipeline in local environment. This creates a pipeline defined in `pipeline/local_pipeline.py`
- `kubeflow_runner.py`: a runner to run pipeline in GCP(Vertex AI) environment. This creates a pipeline defined in `pipeline/pipeline.py`

Sorry about the annoying `Lint` Github Action. I will remove it after this PR

### CODE REVIEW

**Comment by deep-diver** [comment]
@sayakpaul 

reflected the changes in TFRecords. However, the OOM problem of `ImportExampleGen` still persists on Vertex AI Pipeline when using all of the TFRecords. Not sure what is the limit.

**Comment by sayakpaul** [comment]
Will try with 224 resolution then.

File: training_pipeline/models/model.py [hunk 1]
```diff
@@ -1,266 +1,287 @@
...
-            tf.TensorSpec(
-                shape=[None, 224, 224, 3],
-                dtype=tf.float32,
-                name=_transformed_name(_IMAGE_KEY),
-            )
-        )
+def _serving_preprocess(string_input):
+    decoded_input = tf.io.decode_base64(string_input)
+    decoded = tf.io.decode_jpeg(decoded_input, channels=3)
+    decoded = decoded / 255
```
Comment (sayakpaul) [inline training_pipeline/models/model.py:30]: The preprocessing utilities are slightly different from what was used in the TFRecord creation script.

File: training_pipeline/models/model.py [hunk 2]
```diff
@@ -1,266 +1,287 @@
...
+    labels = tf.expand_dims(
+        example_batch["labels"], -1
+    )  # Adds extra dimension, otherwise tf.image.resize won't work.
+    labels = tf.transpose(
+        labels, perm=[0, 1, 2, 3]
+    )  # So, that TF can evaluation the shapes.
 
-    return serve_image_fn
+    images = tf.image.resize(images, _IMAGE_SHAPE)
+    labels = tf.image.resize(labels, _IMAGE_SHAPE)
```
Comment (sayakpaul) [inline training_pipeline/models/model.py:92]: As per [this notebook](https://github.com/deep-diver/semantic-segmentation-ml-pipeline/blob/main/notebooks/parse_tfrecords.ipynb), the labels should be of shape `(batch_size, height, width)`.

This means you need to first add a channel dimension to the labels (segmentation maps), then resize it, and then squeeze the channel dimension again. This is what I have done here

https://github.com/deep-diver/semantic-segmentation-ml-pipeline/blob/7a64785b66b144eacbe4a7547e6b310f46def758/tfrecords/create_tfrecords.py#L52

File: training_pipeline/models/model.py [hunk 3]
```diff
@@ -1,266 +1,287 @@
...
+    images = example_batch["pixel_values"]
+    images = tf.transpose(
+        images, perm=[0, 1, 2, 3]
+    )  # TF can evaluation the shapes (batch_size,  num_channels, height, width)
+    labels = tf.expand_dims(
+        example_batch["labels"], -1
+    )  # Adds extra dimension, otherwise tf.image.resize won't work.
+    labels = tf.transpose(
+        labels, perm=[0, 1, 2, 3]
+    )  # So, that TF can evaluation the shapes.
```
[the remark this answers is not in the document]
Comment (deep-diver) [reply training_pipeline/models/model.py:89]: I have tried removing `tf.transpose` parts, then it didn't work by complaining it does not know the `shape` when performing `resize`

File: training_pipeline/models/model.py (hunk 2 above)
Comment (deep-diver) [reply training_pipeline/models/model.py:92]: this is a **_serving_**preprocess function. Do we need to handle `labels` here too?

Comment (deep-diver) [reply training_pipeline/models/model.py:92]: seems like your explanation is reflected in the code except `transpose`. For instance, the code is doing `expand_dims` => `resize` => `squeeze`.

However, the `resize didn't work without `transpose` by complaining the Tensor doesnt know/have `shape`

File: training_pipeline/models/model.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sayakpaul) [reply training_pipeline/models/model.py:89]: It didn't complain about that in the TFRecord parsing notebook. So, I am not too sure.

File: training_pipeline/models/model.py (hunk 1 above)
Comment (deep-diver) [reply training_pipeline/models/model.py:30]: meaning first `normalize` then `resize`?

Comment (sayakpaul) [reply training_pipeline/models/model.py:30]: You missed out subtracting -1 from the pixel values. Please take a look again.

File: training_pipeline/models/model.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (deep-diver) [reply training_pipeline/models/model.py:89]: will check again then :)

File: training_pipeline/models/model.py (hunk 2 above)
Comment (sayakpaul) [reply training_pipeline/models/model.py:92]: If this utility is a part of the input data pipeline then this problem shouldn't arise as we verified an input pipeline in the TFRecord parsing notebook.

File: training_pipeline/models/model.py (hunk 1 above)
Comment (deep-diver) [reply training_pipeline/models/model.py:30]: hm `label -= 1` is for label data. But, since this is a preprocessing function for the serving, we don't have label data. Or do I miss something?

Comment (sayakpaul) [reply training_pipeline/models/model.py:30]: Oh, then you're right. It's very hard to review 21 changed files in a single PR honestly :D

File: training_pipeline/models/model.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (deep-diver) [reply training_pipeline/models/model.py:89]: Could you take a look at this notebook?
- Note: somehow `tfrec` files are not downloadable from GCS, so I created a very small one with your script, then uploaded to the Colab. If you need those files, please let me know.

https://colab.research.google.com/drive/1hB6pp2vgRytkcwvOrY8PWOj0fkVoilqM?usp=sharing

[the remark this answers is not in the document]
Comment (sayakpaul) [reply training_pipeline/models/model.py:89]: Why can't you directly access the TFRecord files? It's publicly readable:

image

Also, TensorFlow is able to load these files from a publicly readable GCS bucket directly. Could you try again in the notebook?

[the remark this answers is not in the document]
Comment (sayakpaul) [reply training_pipeline/models/model.py:89]: Also, the notebook contains a whole bunch of things that we don't need to isolate the issue. Going forward please, provide a minimal notebook that helps to reproduce the issue.

[the remark this answers is not in the document]
Comment (deep-diver) [reply training_pipeline/models/model.py:89]: sorry my bad!

here is the other notebook built on top of `parse-tfrecords` notebook:
https://colab.research.google.com/drive/1meAImdZUR6NpDDK2fbEichwfTohd9Y7z?usp=sharing

[the remark this answers is not in the document]
Comment (sayakpaul) [reply training_pipeline/models/model.py:89]: Updated the notebook. Looks like some form of pre-evaluation for shape is needed.

**Comment by deep-diver** [comment]
@sayakpaul 

comments are addressed. however, `ImportExampleGen` could not access the GCS buckets in Colab environment.  so, `tfx_pipeline.ipynb` notebook is not modified to use remove GCS buckets. 

Below is the error when running with GCS bucket
```
...
  File "/usr/local/lib/python3.7/dist-packages/google/auth/compute_engine/credentials.py", line 117, in refresh
```

But, I have verified it works with TFX CLI on local and vertex environment (even though it runs into OOM error for now). 

Please let me know, when the resolution of the input data is reduced and uploaded
